m4/libmesh_method.m4

```diff
@@ -40,17 +40,17 @@ AC_DEFUN([LIBMESH_SET_METHODS],
  build_oprof=no
 
  dnl define compiler flags for all methods
- CPPFLAGS_OPT="$CPPFLAGS_OPT $libmesh_CPPFLAGS"
- CXXFLAGS_OPT="$CXXFLAGS_OPT $libmesh_CXXFLAGS"
- CFLAGS_OPT="$CFLAGS_OPT $libmesh_CFLAGS"
+ CPPFLAGS_OPT="$CPPFLAGS_OPT $libmesh_CPPFLAGS $CPPFLAGS"
+ CXXFLAGS_OPT="$CXXFLAGS_OPT $libmesh_CXXFLAGS $CXXFLAGS"
```

Comment:
In the past, we have always purposely avoided appending the `CXXFLAGS` environment variable to the list of flags used by libmesh, under the assumption that the user's `CXXFLAGS` might not work/be correct for libmesh... it could be time to revisit this decision, however. Anyone who is familiar with autoconf-based build systems is going to be used to using `CXXFLAGS` for this purpose, and it's more likely we are just confusing them than preventing errors in dubious environments...

Reply:
https://sourceforge.net/p/libmesh/mailman/message/8105557/ - we made that change more than a decade ago, and now I don't recall why.

I'd personally be in favor of changing it back, but if so let's do it carefully; I think GRINS and probably other packages set libmesh_CXXFLAGS etc. and we'll need to give everyone time to get the non-prefixed counterparts consistent.

Reply:
Yes, I agree it's a pretty big change.  It was meant to solve a particular issue where the `-std=c++11` flag was auto-detected for PGI but was not included in some conftests.  The issue came from this series of events:

1. `-std=c++11` is corretly auto-detected and added to `CXXFLAGS`
https://github.com/libMesh/libmesh/blob/c2501bfbb26e7da832aa9fbea63ff7b27b1b3a39/m4/ax_cxx_compile_stdcxx_11.m4#L124-L150
2. In `libmesh_method.m4`, `CXXFLAGS_DEVEL` is set such that it does *not* include `CXXFLAGS`.  Hence, `CXXFLAGS_DEVEL` does not include `-std=c++11`.
  https://github.com/libMesh/libmesh/blob/c2501bfbb26e7da832aa9fbea63ff7b27b1b3a39/m4/libmesh_method.m4#L52
3. Finally, in `libmesh_compiler_features.m4`, the conftests are run *after* `CXXFLAGS="$CXXFLAGS_DEVEL"`.  This means that, for these conftests, `CXXFLAGS` will not include `-std=c++11`.  At this point, conftests for C++11 features will fail.  
  https://github.com/libMesh/libmesh/blob/c2501bfbb26e7da832aa9fbea63ff7b27b1b3a39/m4/libmesh_compiler_features.m4#L12

Reply:
Here's a possible alternative to adding `CXXFLAGS` to the libmesh flags.   You might be able to add `-std=c++11` to `libmesh_CXXFLAGS` instead of (or in addition to) `CXXFLAGS`.  That should ensure it gets used where necessary.

Reply:
I'm still not sure I understand how `-std=c++11` does not make it into the tests. From what I can tell, there are 4 places where `CXXFLAGS_DEVEL` is changed:

1. configure.ac, line 174: AX_CXX_COMPILE_STDCXX_11() (ax_cxx_compile_stdcxx_11.m4)
2. configure.ac, line 182: LIBMESH_SET_CXX_FLAGS                    (compiler.m4)
3. configure.ac, line 195: LIBMESH_SET_METHODS                      (libmesh_method.m4)
4. configure.ac, line 343: LIBMESH_COMPILER_FEATURES                (libmesh_compiler_features.m4)

At each step, the following happens:

1. Note: `$switch` here is equal to `-std=c++11`
```
CXXFLAGS_DEVEL="$CXXFLAGS_DEVEL $switch"
```

2. GCC-specific flags (for example) are appended
```
CXXFLAGS_DEVEL="$CXXFLAGS_DEVEL -O2 -felide-constructors -g -pedantic -W -Wall -Wextra -Wno-long-long -Wunused"
CXXFLAGS_DEVEL="$CXXFLAGS_DEVEL -Wpointer-arith -Wformat -Wparentheses -Wuninitialized -funroll-loops -fstrict-aliasing -Woverloaded-virtual -Wdisabled-optimization"
```

3. User-defined flags get appended:
```
CXXFLAGS_DEVEL="$CXXFLAGS_DEVEL $libmesh_CXXFLAGS"
```

4. We temporarily set the `CXXFLAGS` for the `ACX_BEST_UNORDERED_MULTIMAP` and friends tests. 
```
CXXFLAGS="$CXXFLAGS_DEVEL"
```

Since `-std=c++11` was appended in step 1 and the `CXXFLAGS_DEVEL` was subsequently only appended to, it should be getting passed to the unordered tests in step 4. Indeed, on my system's config.log I see:

```
configure:29362: checking whether the compiler supports std::unordered_multimap
configure:29390: mpicxx -c  -std=gnu++11 -O2 -felide-constructors -g -pedantic -W -Wall -Wextra -Wno-long-long -Wunused-parameter -Wunused -Wpointer-arith -Wformat -Wparentheses -Wuninitialized -Qunused-arguments -Woverloaded-virtual -fno-limit-debug-info          conftest.cpp >&5
```
As you can see, the `-std=gnu++11` flag is the first one there, as expected.

I'm not sure what in this series of events is breaking for you, and unfortunately, dropbox is blocked for us here, so I haven't had a chance to look at your config.log file yet. I meant to do it from home this weekend, but forgot. Will try and look again tonight.

Reply:
How should I point you to my config.log?  I wanted to post it here somehow, but I'd be happy to just email you.

Reply:
Email would be fine.